### DISCUSSION TITLE
copilot-chat--ask: Wrong type argument: number-or-marker-p, nil

### DISCUSSION THREAD

**Discussion by eamonnsullivan** [chep/copilot-chat.el#31] [Support] [answered]
Hi, I'm new to Copilot and an Emacs user. I'm getting this error when try to send any prompts. I seem to login OK (I hit enter, open a browser, paste in the code and get an "all set" message or similar). I'm using straight.el on: 
```
GNU Emacs 29.4 (build 2, aarch64-apple-darwin23.5.0, NS appkit-2487.60
 Version 14.5 (Build 23F79)) of 2024-07-04
```
My configuration is as follows:
```
(use-package copilot-chat
  :straight (:host github :repo "chep/copilot-chat.el" :files ("*.el"))
  :after (request org markdown-mode))
```

The other package I'm using seems to work as expected (i.e., I get offered code snippets):

```
(use-package copilot
  :straight (:host github :repo "copilot-emacs/copilot.el" :files ("*.el"))
  :bind (("C-c C-c" . copilot-accept-completion))
  :ensure t)
```

What else could I do to try and debug this? I suspect the `nil` here might be the login token, possibly? Does that mean something isn't where it is expected?

**Comment by eamonnsullivan** [comment]
I thought just occurred to me: Could the problem be that I'm using Copilot Business?

**Reply by kassick** [reply]
> Could the problem be that I'm using Copilot Business?

I'm using business too and I don't get errors when using the chat.

**Reply by kassick** [reply]
Can you enable debug-on-error (`M-x toggle-debug-on-error`) and paste the backtrace here?

Also, could you describe how you triggered the error?

**Reply by eamonnsullivan** [reply]
Trace: 
```
Debugger entered--Lisp error: (wrong-type-argument number-or-marker-p nil)
  copilot-chat--auth()
  copilot-chat--ask("help" copilot-chat-prompt-cb)
  copilot-chat-prompt-send()
  funcall-interactively(copilot-chat-prompt-send)
  command-execute(copilot-chat-prompt-send)

```
(But see below -- there seems to be a problem with my access token.)

Steps to reproduce:
1. M-x copilot-chat-display
2. type 'help' then C-c RET

**Comment by chep** [comment]
Are the files `~/.cache/copilot-chat/token` and `~/.config/copilot-chat/github-token` consistent ?

`token` should be json data and `github-token` something like "ghu_something".

**Reply by eamonnsullivan** [reply]
I see the issue, I think. The `~/.cache/copilot-chat/token` is 
```
{"message":"Resource not accessible by personal access token","documentation_url":"https://docs.github.com/rest","status":"403"}
```

**Reply by eamonnsullivan** [reply]
It looks like my fine-grained access token doesn't include copilot chat. I'll generate another and check.

**Reply by eamonnsullivan** [reply]
I created a new access token and explicitly enabled copilot-chat, but I still get the same json in `~/.cache/copilot-chat/token`. I'll try a classic token.

**Reply by chep** [reply]
Copilot chat does not use your tokens, it generates a new one in `~/.config/copilot-chat/github-token`.
Does this file look ok (don't paste it here) ?

**Reply by chep** [reply]
You can try to delete it and restart authentication process. (delete file, then `M-x copilot-chat-reset` and `M-x copilot-chat-display`)

**Reply by eamonnsullivan** [reply]
It does look OK, but I'll try a reset. I've got a series of meetings now (I'm in London) and will try a little later today. Good to know about the access token! I would have kept fiddling with that.

**Reply by eamonnsullivan** [reply]
I've tried a reset and it's working now on the Mac (:champagne: ). Alas, it's not working on my other laptop (Ubuntu 24.04, with the same version of emacs). The `~/.config/copilot-chat/github-token` file is just not being created and the `~/.cache/copilot-chat/token` file has that same message about a 403.

**Reply by chep** [reply]
Do you see any error message during authentication process ? You can check the `*Messages*` buffer for this.

**Reply by eamonnsullivan** [reply]
I don't see any error messages, other than the one in the title. I suspect this might be because I'm using emacs via the snapstore on Ubuntu and maybe it lacks a permission? I'd be surprised, since copilot-emacs seems to work, but it's possible.

**Reply by eamonnsullivan** [reply]
Is there any way to force a relogin? I've removed both `~/.cache/copilot-chat` and `~/.config/copilot-chat` (which actually didn't exist), ran M-x copilot-chat-reset, then M-x copilot-chat-display, typed "help", then C-c RET, and ... nothing. Nothing in *Messages*, no prompts to "Hit enter to open a the browser". Just nothing. There's a *copilot-chat-curl-stderr* buffer, but it just contains: 
```
Process copilot-chat-curl stderr finished
```

**Reply by eamonnsullivan** [reply]
The value of `copilot-chat-curl-program` is "curl". 
```
$ which curl
/usr/bin/curl
$
```

**Reply by chep** [reply]
> Is there any way to force a relogin? I've removed both `~/.cache/copilot-chat` and `~/.config/copilot-chat` (which actually didn't exist), ran M-x copilot-chat-reset, then M-x copilot-chat-display, typed "help", then C-c RET, and ... nothing. Nothing in _Messages_, no prompts to "Hit enter to open a the browser". Just nothing. There's a _copilot-chat-curl-stderr_ buffer, but it just contains:
> 
> ```
> Process copilot-chat-curl stderr finished
> ```

You did force the relogin when removing the directories and calling reset…
That's very strange.
Can you give me the result of `describe-variable` `copilot-chat--instance` ? (don't forget to replace any displayed token by garbage)

**Reply by eamonnsullivan** [reply]
```
copilot-chat--instance is a variable defined in ‘copilot-chat-common.el’.

Its value is
#s(copilot-chat :ready t :github-token "REDACTED" :token
                ((message . "Resource not accessible by personal access token")
                 (documentation_url . "https://docs.github.com/rest")
                 (status . "403"))
                :sessionid "305c184f-ab23-4473c-88b3-667caff1a1371730299989000" :machineid "fbd30dc6ae8ba32192b793ad17c3578cc0ea4951affae1c1a115e16b2c87f5df5" :history
                (("help" "user"))
                :buffers nil)
```
I notice that the redacted token is my own personal github access token, the one I created earlier today, with copilot ticked.

**Reply by eamonnsullivan** [reply]
Ah, is it [reading it from the environment](https://github.com/chep/copilot-chat.el/blob/master/copilot-chat-copilot.el#L83C3-L83C30)? I do have that variable set, with my personal (possibly inadequate) access token.

**Reply by eamonnsullivan** [reply]
Unsetting `GITHUB_TOKEN` does indeed result in a proper login, and I get a new token generated (in `~/.config/copilot-chat/github-token`), but I get the same message and 403 status.

**Reply by eamonnsullivan** [reply]
I'll try again a little later, as I think there might be some caching involved. It just suddenly worked on the Mac after a few tries, so maybe the same will happen on Ubuntu.

**Reply by chep** [reply, marked as answer]
Yes you are right, the GITHUB_TOKEN  variable is used if set. I should remove it, as we can see it can be confusing.

Remove all files again, call reset or restart emacs and it should work.

**Reply by eamonnsullivan** [reply]
It did work. Thank you for the help! 

I do need that GITHUB_TOKEN for other programmes, so it would be good we didn't check that variable. But now I know how to fix the issue.

**Reply by chep** [reply]
I removed the use of the variable.
